### DISCUSSION TITLE
I/O from Midi controllers

### DISCUSSION THREAD

**Discussion by maks** [FundamentalFrequency/.github#58] [Ideas]
Hi @stonepreston , I asked this in a comment a while back on the Reddit thread but would probably be better asked here I think:

Given all the controls are via Midi already, I wonder if it would be possible to change the mapping to make use of say a Novation Launchkey Mini Mk3? which has lots of dials (though not encoders) buttons, keys, modwheel strip etc.

Or even **better** a Akai Fire which has itself 4 encoders, inc 1 push-button encoder, as well as lots of buttons and pads with controllable LEDs and even a 128x64 OLED that can be displayed to using bitmaps via Midi Sysex, though admittedly it would be hard to translate your nice colour UI to its small, monochrome size 🙂 

This might a way to open the project up to more people (like myself) who may already have a midicontroller like that so then would only need to setup the RPI & LCD screen...

Would you have any pointers for me on where to start trying this out with the DAWs source?

**Comment by stonepreston** [comment]
The midi mapping is defined [here](https://github.com/FundamentalFrequency/LMN-3-DAW/blob/ccaa0a7f3ad7c33ba91c7d2d96cd47f111817043/Source/Modules/app_services/MidiCommandManager/MidiCommandManager.h#L171). 

Using a midi controller that does not have encoders but only has pots would be challenging. But you might be able to map it such that turning the pot all the way up is a clockwise encoder turn, and turning the pot all the way down is a counter clockwise encoder turn.

**Reply by maks** [reply]
Brilliant, thanks! That looks very straight forward 👍🏻  And yes I think I'll just try  to first get this working with an Akai Fire which had 4 encoders and bunch of buttons. Actually looking through the button list, though I have a feeling I'll need to have some sort of mapper helper as I'd likely end up using Shift and Alt key combinations on the Akai fire and try mapping the pads into a chromatic keyboard layout or into different scales.

**Comment by Pastitas** [comment]
This could be integrated in the current board via trs-midi connected to one of the uart ports on the teensy 4.1. One issue could be merging the midi messages from outside and the ones generated by the teensy. Another option is to use the free UART port in the rpi.
